<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 üå± Tumor Microenviroment Analysis | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</title>
  <meta name="description" content="Chapter 3 üå± Tumor Microenviroment Analysis | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 üå± Tumor Microenviroment Analysis | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  <meta property="og:type" content="book" />
  
  
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 üå± Tumor Microenviroment Analysis | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  
  
  

<meta name="author" content="Yihao Chen, Li-Na He, Yuanzhe Zhang, Yuelong Shu, Di Zhang, Guangchuang Yu, Zhixiang Zuo" />


<meta name="date" content="2024-06-14" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="biomarker-evaluation.html"/>
<link rel="next" href="prediction-model-construction.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">tigeR documentation</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>üìñ <strong>Getting start with tigeR</strong></a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#introduction"><i class="fa fa-check"></i><b>0.1</b> üî¨ Introduction</a></li>
<li class="chapter" data-level="0.2" data-path="index.html"><a href="index.html#the-workflow-of-tiger"><i class="fa fa-check"></i><b>0.2</b> üèû The workflow of tigeR</a></li>
<li class="chapter" data-level="0.3" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>0.3</b> üõ†Ô∏è Installation</a></li>
<li class="chapter" data-level="0.4" data-path="index.html"><a href="index.html#function-overview"><i class="fa fa-check"></i><b>0.4</b> üó∫Ô∏è Function Overview</a></li>
<li class="chapter" data-level="0.5" data-path="index.html"><a href="index.html#troubleshooting"><i class="fa fa-check"></i><b>0.5</b> üí• Troubleshooting</a></li>
</ul></li>
<li class="chapter" data-level="1" data-path="load-data-into-r-environment.html"><a href="load-data-into-r-environment.html"><i class="fa fa-check"></i><b>1</b> üì• Load data into R environment</a>
<ul>
<li class="chapter" data-level="1.1" data-path="load-data-into-r-environment.html"><a href="load-data-into-r-environment.html#organization-of-built-in-data"><i class="fa fa-check"></i><b>1.1</b> Organization of built-in data</a></li>
<li class="chapter" data-level="1.2" data-path="load-data-into-r-environment.html"><a href="load-data-into-r-environment.html#obtain-data-from-tiger-web-server"><i class="fa fa-check"></i><b>1.2</b> Obtain data from tigeR web server</a></li>
<li class="chapter" data-level="1.3" data-path="load-data-into-r-environment.html"><a href="load-data-into-r-environment.html#obtain-data-from-experimenthub"><i class="fa fa-check"></i><b>1.3</b> Obtain data from ExperimentHub</a></li>
<li class="chapter" data-level="1.4" data-path="load-data-into-r-environment.html"><a href="load-data-into-r-environment.html#pre-processing-of-custom-data"><i class="fa fa-check"></i><b>1.4</b> Pre-processing of Custom Data</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html"><i class="fa fa-check"></i><b>2</b> üö© Biomarker Evaluation</a>
<ul>
<li class="chapter" data-level="2.1" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#integrate-analysis"><i class="fa fa-check"></i><b>2.1</b> Integrate analysis</a></li>
<li class="chapter" data-level="2.2" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#differential-analysis"><i class="fa fa-check"></i><b>2.2</b> Differential analysis</a></li>
<li class="chapter" data-level="2.3" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#suvival-analysis"><i class="fa fa-check"></i><b>2.3</b> Suvival analysis</a></li>
<li class="chapter" data-level="2.4" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#calculate-comprehensive-signature-score"><i class="fa fa-check"></i><b>2.4</b> Calculate comprehensive signature score</a></li>
<li class="chapter" data-level="2.5" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#assess-the-performance-of-signature"><i class="fa fa-check"></i><b>2.5</b> Assess the Performance of Signature</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html"><i class="fa fa-check"></i><b>3</b> üå± Tumor Microenviroment Analysis</a>
<ul>
<li class="chapter" data-level="3.1" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#availiable-tme-analysis-method-in-tiger"><i class="fa fa-check"></i><b>3.1</b> Availiable TME Analysis Method in tigeR</a></li>
<li class="chapter" data-level="3.2" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#derive-the-proportions-of-different-tme-cell-types"><i class="fa fa-check"></i><b>3.2</b> Derive the Proportions of Different TME Cell Types</a></li>
<li class="chapter" data-level="3.3" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#visualization-and-comparing-the-cell-proportions"><i class="fa fa-check"></i><b>3.3</b> Visualization and Comparing the Cell Proportions</a></li>
<li class="chapter" data-level="3.4" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#searching-for-key-cell-types-associated-with-immunotherapy-response"><i class="fa fa-check"></i><b>3.4</b> Searching for Key Cell Types Associated with Immunotherapy Response</a></li>
<li class="chapter" data-level="3.5" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#refining-a-custom-reference-matrix-for-tme-deconvolution-from-single-cell-sequencing-data"><i class="fa fa-check"></i><b>3.5</b> Refining a Custom Reference Matrix for TME Deconvolution from Single-cell Sequencing Data</a></li>
<li class="chapter" data-level="" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#more-details-about-tme-analysis"><i class="fa fa-check"></i>üìù More Details About TME Analysis</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html"><i class="fa fa-check"></i><b>4</b> ü§ñ Prediction Model Construction</a>
<ul>
<li class="chapter" data-level="4.1" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#build-model"><i class="fa fa-check"></i><b>4.1</b> Build Model</a></li>
<li class="chapter" data-level="4.2" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#test-model"><i class="fa fa-check"></i><b>4.2</b> Test Model</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="response-prediction.html"><a href="response-prediction.html"><i class="fa fa-check"></i><b>5</b> üîÆ Response Prediction</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="tumor-microenviroment-analysis" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> üå± Tumor Microenviroment Analysis<a href="#tumor-microenviroment-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="availiable-tme-analysis-method-in-tiger" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Availiable TME Analysis Method in tigeR<a href="tumor-microenviroment-analysis.html#availiable-tme-analysis-method-in-tiger" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>10 open source tumor microenvironment deconvolution methods are indlcued in tigeR.</p>
<div style="width:780px; height:200px; overflow-y: scroll; overflow-x: hidden;">
<table>
<colgroup>
<col width="25%" />
<col width="36%" />
<col width="20%" />
<col width="18%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">Algorithm</th>
<th align="center">license</th>
<th align="center">PMID</th>
<th align="center">allow custom signature matrix</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center"><a href="http://cistrome.org/TIMER/">TIMER</a></td>
<td align="center">free (GPL 2.0)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/27549193/">27549193</a></td>
<td align="center">‚àö</td>
</tr>
<tr class="even">
<td align="center"><a href="https://cibersort.stanford.edu/">CIBERSORT</a></td>
<td align="center">free for non-commercial use only</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/25822800/">25822800</a></td>
<td align="center">‚àö</td>
</tr>
<tr class="odd">
<td align="center"><a href="https://github.com/ebecht/MCPcounter">MCPCounter</a></td>
<td align="center">free (<a href="https://github.com/ebecht/MCPcounter/blob/master/Source/License">GPL 3.0</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/27765066/">27765066</a></td>
<td align="center">‚àö</td>
</tr>
<tr class="even">
<td align="center"><a href="http://xcell.ucsf.edu/">xCell</a></td>
<td align="center">free (<a href="https://github.com/dviraran/xCell/blob/master/DESCRIPTION">GPL 3.0</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/29141660/">29141660</a></td>
<td align="center">√ó</td>
</tr>
<tr class="odd">
<td align="center"><a href="https://github.com/icbi-lab/Immunophenogram">IPS</a></td>
<td align="center">free (BSD)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/28052254/">28052254</a></td>
<td align="center">√ó</td>
</tr>
<tr class="even">
<td align="center"><a href="https://gfellerlab.shinyapps.io/EPIC_1-1/">EPIC</a></td>
<td align="center">free for non-commercial use only (<a href="https://github.com/GfellerLab/EPIC/blob/master/LICENSE">Academic License</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/29130882/">29130882</a></td>
<td align="center">‚àö</td>
</tr>
<tr class="odd">
<td align="center"><a href="https://gfellerlab.shinyapps.io/EPIC_1-1/">ESTIMATE</a></td>
<td align="center">free (<a href="https://bioinformatics.mdanderson.org/public-software/estimate/">GPL 2.0</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/24113773/">24113773</a></td>
<td align="center">√ó</td>
</tr>
<tr class="even">
<td align="center"><a href="https://giannimonaco.shinyapps.io/ABIS/">ABIS</a></td>
<td align="center">free (<a href="https://github.com/giannimonaco/ABIS">GPL 2.0</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/30726743/">30726743</a></td>
<td align="center">‚àö</td>
</tr>
<tr class="odd">
<td align="center"><a href="https://olliecast.shinyapps.io/Deconvolution_Benchmarking/">ConsensusTME</a></td>
<td align="center">free (<a href="https://github.com/cansysbio/ConsensusTME/blob/master/LICENSE.md">GPL 3.0</a>)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/31641033/">31641033</a></td>
<td align="center">√ó</td>
</tr>
<tr class="even">
<td align="center"><a href="http://icbi.at/software/quantiseq/doc/index.html">quanTIseq</a></td>
<td align="center">free (BSD)</td>
<td align="center"><a href="https://pubmed.ncbi.nlm.nih.gov/31126321/">31126321</a></td>
<td align="center">‚àö</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="derive-the-proportions-of-different-tme-cell-types" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Derive the Proportions of Different TME Cell Types<a href="tumor-microenviroment-analysis.html#derive-the-proportions-of-different-tme-cell-types" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>‚ÄÉYou can use the function <strong>deconv_TME()</strong> to derive the proportions of different tumor microenvironment cell types from gene expression data with these tools.</p>
<pre><code>devtools::install_github(&#39;dviraran/xCell&#39;)
devtools::install_github(&quot;GfellerLab/EPIC&quot;)
devtools::install_github(&quot;cansysbio/ConsensusTME&quot;)
devtools::install_github(&quot;federicomarini/quantiseqr&quot;)</code></pre>
<div style="width:780px; height:200px; overflow-y: scroll; overflow-x: hidden;">
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="tumor-microenviroment-analysis.html#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="do">## TIMER</span></span>
<span id="cb15-2"><a href="tumor-microenviroment-analysis.html#cb15-2" aria-hidden="true" tabindex="-1"></a>frac1 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;TIMER&quot;</span>)</span></code></pre></div>
<pre><code>## Found 130 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="tumor-microenviroment-analysis.html#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="do">## CIBERSORT</span></span>
<span id="cb17-2"><a href="tumor-microenviroment-analysis.html#cb17-2" aria-hidden="true" tabindex="-1"></a>frac2 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;CIBERSORT&quot;</span>)</span></code></pre></div>
<pre><code>## Called from: CIBERSORT(SE = SE, ...)
## debug: isList &lt;- is.list(SE)
## debug: exp_mtr &lt;- bind_mtr(SE, isList)
## debug: if (missing(sig_matrix)) {
##     LM22 &lt;- NULL
##     data(LM22, package = &quot;tigeR&quot;, envir = current_env())
##     sig_matrix &lt;- LM22
##     result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
##     TME_data &lt;- as.data.frame(result[, 1:22])
## } else {
##     result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
##     TME_data &lt;- as.data.frame(result)
## }
## debug: LM22 &lt;- NULL
## debug: data(LM22, package = &quot;tigeR&quot;, envir = current_env())
## debug: sig_matrix &lt;- LM22
## debug: result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
## debug: TME_data &lt;- as.data.frame(result[, 1:22])
## debug: TME_data$group &lt;- bind_meta(SE, isList)$response_NR
## debug: TME_data$sample &lt;- rownames(TME_data)
## debug: TME_New &lt;- reshape2::melt(TME_data)
## debug: colnames(TME_New) &lt;- c(&quot;Group&quot;, &quot;Sample&quot;, &quot;Celltype&quot;, &quot;Composition&quot;)
## debug: plot_order &lt;- TME_New[TME_New$Group == &quot;R&quot;, ] %&gt;% dplyr::group_by(.data$Celltype) %&gt;% 
##     dplyr::summarise(m = stats::median(.data$Composition)) %&gt;% 
##     dplyr::arrange(dplyr::desc(.data$m)) %&gt;% dplyr::pull(.data$Celltype)
## debug: TME_New$Celltype &lt;- factor(TME_New$Celltype, levels = plot_order)
## debug: TME_New &lt;- TME_New[!TME_New$Group == &quot;UNK&quot;, ]
## debug: rs &lt;- c()
## debug: warning_status &lt;- 0
## debug: for (i in levels(TME_New$Celltype)) {
##     m &lt;- TME_New[TME_New$Celltype == i, ]
##     R_S &lt;- m[m$Group == &quot;R&quot;, 4]
##     N_S &lt;- m[m$Group == &quot;N&quot;, 4]
##     if (any(R_S %in% N_S)) 
##         warning_status &lt;- 1
##     rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## }
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: if (warning_status) message(&quot;There are identical relative abundance values in groups R and N for the &#39;&quot;, 
##     i, &quot;&#39;. The p value of the Wilcoxon signed-rank test may not be precise due to ties in the data.&quot;)
## debug: message(&quot;There are identical relative abundance values in groups R and N for the &#39;&quot;, 
##     i, &quot;&#39;. The p value of the Wilcoxon signed-rank test may not be precise due to ties in the data.&quot;)
## debug: selected_cells &lt;- levels(TME_New$Celltype)
## debug: if (length(selected_cells) &lt; 5) {
##     names(rs) &lt;- seq_along(rs)
##     selected_cells &lt;- levels(TME_New$Celltype)[as.numeric(names(sort(rs)[1:5]))]
## }
## debug: ciber_theme &lt;- ggplot2::theme(plot.title = element_text(size = 12, 
##     color = &quot;black&quot;, hjust = 0.5), axis.title = element_text(size = 10, 
##     color = &quot;black&quot;), axis.text = element_text(size = 10, color = &quot;black&quot;), 
##     axis.text.x = element_text(angle = 45, hjust = 1), legend.position = &quot;top&quot;, 
##     legend.text = element_text(size = 12), legend.title = element_text(size = 12))
## debug: if (style == &quot;raw&quot;) {
##     box_TME &lt;- ggplot2::ggplot(TME_New[TME_New$Celltype %in% 
##         selected_cells, ], aes(x = .data$Celltype, y = .data$Composition)) + 
##         ggplot2::labs(y = &quot;Cell composition&quot;, x = NULL, title = &quot;TME Cell composition&quot;) + 
##         ggplot2::geom_boxplot(aes(fill = .data$Group), position = position_dodge(0.5), 
##             width = 0.5, outlier.alpha = 0) + ggplot2::scale_fill_manual(values = group_color) + 
##         ggplot2::theme_classic() + ciber_theme
##     y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
##     box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##         label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = TRUE, 
##         label.y = y_max * 1.1) + coord_cartesian(ylim = c(0, 
##         y_max * 1.1))
## }
## debug: if (style == &quot;elegant&quot;) {
##     box_TME &lt;- ggplot(TME_New, aes(x = .data$Celltype, y = .data$Composition, 
##         fill = .data$Group)) + stat_boxplot(data = TME_New, geom = &quot;errorbar&quot;, 
##         width = 1, color = &quot;black&quot;, linetype = &quot;solid&quot;, position = position_dodge(0.8), 
##         linewidth = 0.7) + stat_boxplot(geom = &quot;boxplot&quot;, color = &quot;black&quot;, 
##         linetype = &quot;solid&quot;, position = position_dodge(0.8), linewidth = 0.7, 
##         width = 0.8, outlier.shape = 19) + ggpubr::theme_classic2() + 
##         ciber_theme + scale_fill_manual(values = group_color)
##     y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
##     box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##         label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = FALSE, 
##         label.y = y_max * 1.1, size = 3) + coord_cartesian(ylim = c(0, 
##         y_max * 1.1))
## }
## debug: box_TME &lt;- ggplot(TME_New, aes(x = .data$Celltype, y = .data$Composition, 
##     fill = .data$Group)) + stat_boxplot(data = TME_New, geom = &quot;errorbar&quot;, 
##     width = 1, color = &quot;black&quot;, linetype = &quot;solid&quot;, position = position_dodge(0.8), 
##     linewidth = 0.7) + stat_boxplot(geom = &quot;boxplot&quot;, color = &quot;black&quot;, 
##     linetype = &quot;solid&quot;, position = position_dodge(0.8), linewidth = 0.7, 
##     width = 0.8, outlier.shape = 19) + ggpubr::theme_classic2() + 
##     ciber_theme + scale_fill_manual(values = group_color)
## debug: y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
## debug: box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##     label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = FALSE, 
##     label.y = y_max * 1.1, size = 3) + coord_cartesian(ylim = c(0, 
##     y_max * 1.1))
## debug: list(t(result), box_TME)</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="tumor-microenviroment-analysis.html#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="do">## MCPCounter</span></span>
<span id="cb19-2"><a href="tumor-microenviroment-analysis.html#cb19-2" aria-hidden="true" tabindex="-1"></a>frac3 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;MCPCounter&quot;</span>)</span>
<span id="cb19-3"><a href="tumor-microenviroment-analysis.html#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="tumor-microenviroment-analysis.html#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="do">## xCell</span></span>
<span id="cb19-5"><a href="tumor-microenviroment-analysis.html#cb19-5" aria-hidden="true" tabindex="-1"></a>frac4 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;xCell&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Num. of genes: 10808&quot;
## Setting parallel calculations through a MulticoreParam back-end
## with workers=4 and tasks=100.
## Estimating ssGSEA scores for 489 gene sets.
## [1] &quot;Calculating ranks...&quot;
## [1] &quot;Calculating absolute values from ranks...&quot;
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   4%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |===============                                                       |  21%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |====================                                                  |  29%
  |                                                                            
  |======================                                                |  32%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |============================                                          |  39%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  61%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |================================================                      |  68%
  |                                                                            
  |==================================================                    |  71%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |=======================================================               |  79%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |====================================================================  |  96%
  |                                                                            
  |======================================================================| 100%</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="tumor-microenviroment-analysis.html#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="do">## IPS</span></span>
<span id="cb21-2"><a href="tumor-microenviroment-analysis.html#cb21-2" aria-hidden="true" tabindex="-1"></a>frac5 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;IPS&quot;</span>)</span>
<span id="cb21-3"><a href="tumor-microenviroment-analysis.html#cb21-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-4"><a href="tumor-microenviroment-analysis.html#cb21-4" aria-hidden="true" tabindex="-1"></a><span class="do">## EPIC</span></span>
<span id="cb21-5"><a href="tumor-microenviroment-analysis.html#cb21-5" aria-hidden="true" tabindex="-1"></a>frac6 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;epic&quot;</span>)</span>
<span id="cb21-6"><a href="tumor-microenviroment-analysis.html#cb21-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-7"><a href="tumor-microenviroment-analysis.html#cb21-7" aria-hidden="true" tabindex="-1"></a><span class="do">## ESTIMATE</span></span>
<span id="cb21-8"><a href="tumor-microenviroment-analysis.html#cb21-8" aria-hidden="true" tabindex="-1"></a>frac7 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;ESTIMATE&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;Merged dataset includes 10130 genes (282 mismatched).&quot;
## [1] &quot;1 gene set: StromalSignature  overlap= 137&quot;
## [1] &quot;2 gene set: ImmuneSignature  overlap= 141&quot;</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="tumor-microenviroment-analysis.html#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="do">## ABIS</span></span>
<span id="cb23-2"><a href="tumor-microenviroment-analysis.html#cb23-2" aria-hidden="true" tabindex="-1"></a>frac8 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;ABIS&quot;</span>)</span>
<span id="cb23-3"><a href="tumor-microenviroment-analysis.html#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="tumor-microenviroment-analysis.html#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="do">## ConsensusTME</span></span>
<span id="cb23-5"><a href="tumor-microenviroment-analysis.html#cb23-5" aria-hidden="true" tabindex="-1"></a>frac9 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;ConsensusTME&quot;</span>)</span></code></pre></div>
<pre><code>## Producing ConsensusTME Estimates Using The Following Parameters:
##  Statistical Framework: &quot;ssgsea&quot;
##  Gene Sets For Cancer Type: &quot;SKCM&quot;
##  Sample Size: 28
## Estimating ssGSEA scores for 19 gene sets.
## [1] &quot;Calculating ranks...&quot;
## [1] &quot;Calculating absolute values from ranks...&quot;
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |==                                                                    |   4%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |===============                                                       |  21%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |====================                                                  |  29%
  |                                                                            
  |======================                                                |  32%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |============================                                          |  39%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |==========================================                            |  61%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |================================================                      |  68%
  |                                                                            
  |==================================================                    |  71%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |=======================================================               |  79%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |====================================================================  |  96%
  |                                                                            
  |======================================================================| 100%
## 
## [1] &quot;Normalizing...&quot;</code></pre>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="tumor-microenviroment-analysis.html#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="do">## quanTIseq</span></span>
<span id="cb25-2"><a href="tumor-microenviroment-analysis.html#cb25-2" aria-hidden="true" tabindex="-1"></a>frac10 <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE78220,<span class="at">method=</span><span class="st">&quot;quanTIseq&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="visualization-and-comparing-the-cell-proportions" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Visualization and Comparing the Cell Proportions<a href="tumor-microenviroment-analysis.html#visualization-and-comparing-the-cell-proportions" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="tumor-microenviroment-analysis.html#cb26-1" aria-hidden="true" tabindex="-1"></a>cell1 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;T cells CD4&quot;</span>,<span class="st">&quot;Neutrophil&quot;</span>, <span class="st">&quot;Macrophage&quot;</span>,<span class="st">&quot;mDCs&quot;</span>,<span class="st">&quot;B cells&quot;</span>, <span class="st">&quot;T cells CD8&quot;</span>)</span>
<span id="cb26-2"><a href="tumor-microenviroment-analysis.html#cb26-2" aria-hidden="true" tabindex="-1"></a>pie1 <span class="ot">&lt;-</span> <span class="fu">fraction_pie</span>(<span class="fu">cell_name_filter</span>(frac1),<span class="at">feature=</span><span class="fu">factor</span>(cell1, <span class="at">levels =</span> cell1))</span>
<span id="cb26-3"><a href="tumor-microenviroment-analysis.html#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="tumor-microenviroment-analysis.html#cb26-4" aria-hidden="true" tabindex="-1"></a>cell2 <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;DCs resting&quot;</span>, <span class="st">&quot;T cells CD8&quot;</span>, <span class="st">&quot;T cells CD4 naive&quot;</span>, <span class="st">&quot;Macrophages M2&quot;</span>, <span class="st">&quot;Yd T cells&quot;</span>, <span class="st">&quot;Monocytes&quot;</span>,<span class="st">&quot;Mast cells resting&quot;</span>, <span class="st">&quot;Neutrophils&quot;</span>, <span class="st">&quot;Tregs&quot;</span>,<span class="st">&quot;B cells naive&quot;</span>)</span>
<span id="cb26-5"><a href="tumor-microenviroment-analysis.html#cb26-5" aria-hidden="true" tabindex="-1"></a>pie2 <span class="ot">&lt;-</span> <span class="fu">fraction_pie</span>(<span class="fu">cell_name_filter</span>(frac2[[<span class="dv">1</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">22</span>,]),<span class="at">feature=</span><span class="fu">factor</span>(cell2, <span class="at">levels =</span> cell2))</span>
<span id="cb26-6"><a href="tumor-microenviroment-analysis.html#cb26-6" aria-hidden="true" tabindex="-1"></a>pie1</span></code></pre></div>
<p><img src="03-TME-Analysis_files/figure-html/unnamed-chunk-3-1.png" width="2496" /></p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="tumor-microenviroment-analysis.html#cb27-1" aria-hidden="true" tabindex="-1"></a>pie2</span></code></pre></div>
<p><img src="03-TME-Analysis_files/figure-html/unnamed-chunk-3-2.png" width="2496" /></p>
</div>
<div id="searching-for-key-cell-types-associated-with-immunotherapy-response" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Searching for Key Cell Types Associated with Immunotherapy Response<a href="tumor-microenviroment-analysis.html#searching-for-key-cell-types-associated-with-immunotherapy-response" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="tumor-microenviroment-analysis.html#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="do">## TIMER</span></span>
<span id="cb28-2"><a href="tumor-microenviroment-analysis.html#cb28-2" aria-hidden="true" tabindex="-1"></a>TM <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE91061,<span class="at">method =</span> <span class="st">&quot;TIMER&quot;</span>)</span></code></pre></div>
<pre><code>## Found 125 genes with uniform expression within a single batch (all zeros); these will not be adjusted for batch.</code></pre>
<div class="sourceCode" id="cb30"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb30-1"><a href="tumor-microenviroment-analysis.html#cb30-1" aria-hidden="true" tabindex="-1"></a>TM_SE <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(<span class="at">assays=</span><span class="fu">SimpleList</span>(TM),</span>
<span id="cb30-2"><a href="tumor-microenviroment-analysis.html#cb30-2" aria-hidden="true" tabindex="-1"></a>                               <span class="at">colData=</span><span class="fu">colData</span>(MEL_GSE91061))</span>
<span id="cb30-3"><a href="tumor-microenviroment-analysis.html#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="fu">browse_biomk</span>(<span class="at">SE=</span>TM_SE)</span></code></pre></div>
<p><img src="03-TME-Analysis_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
</div>
<div id="refining-a-custom-reference-matrix-for-tme-deconvolution-from-single-cell-sequencing-data" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Refining a Custom Reference Matrix for TME Deconvolution from Single-cell Sequencing Data<a href="tumor-microenviroment-analysis.html#refining-a-custom-reference-matrix-for-tme-deconvolution-from-single-cell-sequencing-data" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="tumor-microenviroment-analysis.html#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span></code></pre></div>
<pre><code>## Loading required package: SeuratObject</code></pre>
<pre><code>## Loading required package: sp</code></pre>
<pre><code>## 
## Attaching package: &#39;sp&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     %over%</code></pre>
<pre><code>## 
## Attaching package: &#39;SeuratObject&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:SummarizedExperiment&#39;:
## 
##     Assays</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomeInfoDb&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following object is masked from &#39;package:IRanges&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following object is masked from &#39;package:S4Vectors&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following object is masked from &#39;package:BiocGenerics&#39;:
## 
##     intersect</code></pre>
<pre><code>## The following object is masked from &#39;package:base&#39;:
## 
##     intersect</code></pre>
<pre><code>## 
## Attaching package: &#39;Seurat&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:SummarizedExperiment&#39;:
## 
##     Assays</code></pre>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb46-1"><a href="tumor-microenviroment-analysis.html#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;magrittr&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:GenomicRanges&#39;:
## 
##     subtract</code></pre>
<div class="sourceCode" id="cb49"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb49-1"><a href="tumor-microenviroment-analysis.html#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GenomicFeatures)</span></code></pre></div>
<pre><code>## Warning: package &#39;GenomicFeatures&#39; was built under R version 4.3.3</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb52-1"><a href="tumor-microenviroment-analysis.html#cb52-1" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,<span class="st">&quot;pbmc.rds&quot;</span>,</span>
<span id="cb52-2"><a href="tumor-microenviroment-analysis.html#cb52-2" aria-hidden="true" tabindex="-1"></a>                            <span class="at">package =</span> <span class="st">&quot;tigeR&quot;</span>,</span>
<span id="cb52-3"><a href="tumor-microenviroment-analysis.html#cb52-3" aria-hidden="true" tabindex="-1"></a>                            <span class="at">mustWork =</span> <span class="cn">TRUE</span>))</span>
<span id="cb52-4"><a href="tumor-microenviroment-analysis.html#cb52-4" aria-hidden="true" tabindex="-1"></a>assay_Seu <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(pbmc, <span class="at">layer =</span> <span class="st">&quot;count&quot;</span>)</span>
<span id="cb52-5"><a href="tumor-microenviroment-analysis.html#cb52-5" aria-hidden="true" tabindex="-1"></a>count <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(assay_Seu)</span>
<span id="cb52-6"><a href="tumor-microenviroment-analysis.html#cb52-6" aria-hidden="true" tabindex="-1"></a>tpm <span class="ot">&lt;-</span> <span class="fu">count2tpm</span>(count)</span>
<span id="cb52-7"><a href="tumor-microenviroment-analysis.html#cb52-7" aria-hidden="true" tabindex="-1"></a>tpm_assay <span class="ot">&lt;-</span> <span class="fu">CreateAssayObject</span>(<span class="at">counts =</span> tpm)</span>
<span id="cb52-8"><a href="tumor-microenviroment-analysis.html#cb52-8" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">&quot;TPM&quot;</span>]] <span class="ot">&lt;-</span> tpm_assay</span>
<span id="cb52-9"><a href="tumor-microenviroment-analysis.html#cb52-9" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(pbmc) <span class="ot">&lt;-</span> <span class="st">&quot;TPM&quot;</span></span>
<span id="cb52-10"><a href="tumor-microenviroment-analysis.html#cb52-10" aria-hidden="true" tabindex="-1"></a>pbmc[[<span class="st">&quot;RNA&quot;</span>]] <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb52-11"><a href="tumor-microenviroment-analysis.html#cb52-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-12"><a href="tumor-microenviroment-analysis.html#cb52-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-13"><a href="tumor-microenviroment-analysis.html#cb52-13" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pbmc, </span>
<span id="cb52-14"><a href="tumor-microenviroment-analysis.html#cb52-14" aria-hidden="true" tabindex="-1"></a>                             <span class="at">selection.method =</span> <span class="st">&quot;vst&quot;</span>, </span>
<span id="cb52-15"><a href="tumor-microenviroment-analysis.html#cb52-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">nfeatures =</span> <span class="dv">2000</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-16"><a href="tumor-microenviroment-analysis.html#cb52-16" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc, </span>
<span id="cb52-17"><a href="tumor-microenviroment-analysis.html#cb52-17" aria-hidden="true" tabindex="-1"></a>                      <span class="at">normalization.method =</span> <span class="st">&quot;LogNormalize&quot;</span>, </span>
<span id="cb52-18"><a href="tumor-microenviroment-analysis.html#cb52-18" aria-hidden="true" tabindex="-1"></a>                      <span class="at">scale.factor =</span> <span class="dv">10000</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-19"><a href="tumor-microenviroment-analysis.html#cb52-19" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc, <span class="at">features =</span>  <span class="fu">rownames</span>(pbmc), <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-20"><a href="tumor-microenviroment-analysis.html#cb52-20" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pbmc, <span class="at">features =</span> <span class="fu">VariableFeatures</span>(<span class="at">object =</span> pbmc), <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-21"><a href="tumor-microenviroment-analysis.html#cb52-21" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(pbmc, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-22"><a href="tumor-microenviroment-analysis.html#cb52-22" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(pbmc, <span class="at">resolution =</span> <span class="fl">0.5</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb52-23"><a href="tumor-microenviroment-analysis.html#cb52-23" aria-hidden="true" tabindex="-1"></a>new.cluster.ids <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Naive_CD4T&quot;</span>, <span class="st">&quot;CD14_Mono&quot;</span>, <span class="st">&quot;Memory_CD4T&quot;</span>, <span class="st">&quot;B_cell&quot;</span>, <span class="st">&quot;CD8T&quot;</span>, <span class="st">&quot;FCGR3A_Mono&quot;</span>, <span class="st">&quot;NK&quot;</span>, <span class="st">&quot;DC&quot;</span>, <span class="st">&quot;Platelet&quot;</span>)</span>
<span id="cb52-24"><a href="tumor-microenviroment-analysis.html#cb52-24" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(new.cluster.ids) <span class="ot">&lt;-</span> <span class="fu">levels</span>(pbmc<span class="sc">$</span>seurat_clusters)</span>
<span id="cb52-25"><a href="tumor-microenviroment-analysis.html#cb52-25" aria-hidden="true" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> <span class="fu">RenameIdents</span>(pbmc, new.cluster.ids)</span>
<span id="cb52-26"><a href="tumor-microenviroment-analysis.html#cb52-26" aria-hidden="true" tabindex="-1"></a><span class="fu">PCAPlot</span>(pbmc, <span class="at">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">pt.size =</span> <span class="fl">0.5</span>) <span class="sc">+</span> <span class="fu">NoLegend</span>()</span></code></pre></div>
<p><img src="03-TME-Analysis_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb53-1"><a href="tumor-microenviroment-analysis.html#cb53-1" aria-hidden="true" tabindex="-1"></a>pbmc<span class="sc">$</span>celltype <span class="ot">&lt;-</span> <span class="fu">Idents</span>(pbmc)</span>
<span id="cb53-2"><a href="tumor-microenviroment-analysis.html#cb53-2" aria-hidden="true" tabindex="-1"></a>ref_sig_mtr <span class="ot">&lt;-</span> <span class="fu">refine_Reference</span>(pbmc)</span></code></pre></div>
<pre><code>## Calculating cluster Naive_CD4T</code></pre>
<pre><code>## For a (much!) faster implementation of the Wilcoxon Rank Sum Test,
## (default method for FindMarkers) please install the presto package
## --------------------------------------------
## install.packages(&#39;devtools&#39;)
## devtools::install_github(&#39;immunogenomics/presto&#39;)
## --------------------------------------------
## After installation of presto, Seurat will automatically use the more 
## efficient implementation (no further action necessary).
## This message will be shown once per session</code></pre>
<pre><code>## Calculating cluster CD14_Mono</code></pre>
<pre><code>## Calculating cluster Memory_CD4T</code></pre>
<pre><code>## Calculating cluster B_cell</code></pre>
<pre><code>## Calculating cluster CD8T</code></pre>
<pre><code>## Calculating cluster FCGR3A_Mono</code></pre>
<pre><code>## Calculating cluster NK</code></pre>
<pre><code>## Calculating cluster DC</code></pre>
<pre><code>## Calculating cluster Platelet</code></pre>
<pre><code>## The following functions and any applicable methods accept the dots: CreateSeuratObject</code></pre>
<pre><code>## As of Seurat v5, we recommend using AggregateExpression to perform pseudo-bulk analysis.
## Names of identity class contain underscores (&#39;_&#39;), replacing with dashes (&#39;-&#39;)
## This message is displayed once per session.</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="tumor-microenviroment-analysis.html#cb66-1" aria-hidden="true" tabindex="-1"></a>cell_fraction <span class="ot">&lt;-</span> <span class="fu">deconv_TME</span>(MEL_GSE100797,<span class="at">method =</span> <span class="st">&quot;CIBERSORT&quot;</span>, <span class="at">sig_matrix=</span>ref_sig_mtr)</span></code></pre></div>
<pre><code>## Called from: CIBERSORT(SE = SE, ...)
## debug: isList &lt;- is.list(SE)
## debug: exp_mtr &lt;- bind_mtr(SE, isList)
## debug: if (missing(sig_matrix)) {
##     LM22 &lt;- NULL
##     data(LM22, package = &quot;tigeR&quot;, envir = current_env())
##     sig_matrix &lt;- LM22
##     result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
##     TME_data &lt;- as.data.frame(result[, 1:22])
## } else {
##     result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
##     TME_data &lt;- as.data.frame(result)
## }
## debug: result &lt;- Ciber(sig_matrix, exp_mtr, perm, QN)
## debug: TME_data &lt;- as.data.frame(result)
## debug: TME_data$group &lt;- bind_meta(SE, isList)$response_NR
## debug: TME_data$sample &lt;- rownames(TME_data)
## debug: TME_New &lt;- reshape2::melt(TME_data)</code></pre>
<pre><code>## Using group, sample as id variables</code></pre>
<pre><code>## debug: colnames(TME_New) &lt;- c(&quot;Group&quot;, &quot;Sample&quot;, &quot;Celltype&quot;, &quot;Composition&quot;)
## debug: plot_order &lt;- TME_New[TME_New$Group == &quot;R&quot;, ] %&gt;% dplyr::group_by(.data$Celltype) %&gt;% 
##     dplyr::summarise(m = stats::median(.data$Composition)) %&gt;% 
##     dplyr::arrange(dplyr::desc(.data$m)) %&gt;% dplyr::pull(.data$Celltype)
## debug: TME_New$Celltype &lt;- factor(TME_New$Celltype, levels = plot_order)
## debug: TME_New &lt;- TME_New[!TME_New$Group == &quot;UNK&quot;, ]
## debug: rs &lt;- c()
## debug: warning_status &lt;- 0
## debug: for (i in levels(TME_New$Celltype)) {
##     m &lt;- TME_New[TME_New$Celltype == i, ]
##     R_S &lt;- m[m$Group == &quot;R&quot;, 4]
##     N_S &lt;- m[m$Group == &quot;N&quot;, 4]
##     if (any(R_S %in% N_S)) 
##         warning_status &lt;- 1
##     rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## }
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: m &lt;- TME_New[TME_New$Celltype == i, ]
## debug: R_S &lt;- m[m$Group == &quot;R&quot;, 4]
## debug: N_S &lt;- m[m$Group == &quot;N&quot;, 4]
## debug: if (any(R_S %in% N_S)) warning_status &lt;- 1
## debug: warning_status &lt;- 1
## debug: rs &lt;- c(rs, suppressWarnings(stats::wilcox.test(R_S, N_S)$p.value))
## debug: if (warning_status) message(&quot;There are identical relative abundance values in groups R and N for the &#39;&quot;, 
##     i, &quot;&#39;. The p value of the Wilcoxon signed-rank test may not be precise due to ties in the data.&quot;)
## debug: message(&quot;There are identical relative abundance values in groups R and N for the &#39;&quot;, 
##     i, &quot;&#39;. The p value of the Wilcoxon signed-rank test may not be precise due to ties in the data.&quot;)</code></pre>
<pre><code>## There are identical relative abundance values in groups R and N for the &#39;Platelet&#39;. The p value of the Wilcoxon signed-rank test may not be precise due to ties in the data.</code></pre>
<pre><code>## debug: selected_cells &lt;- levels(TME_New$Celltype)
## debug: if (length(selected_cells) &lt; 5) {
##     names(rs) &lt;- seq_along(rs)
##     selected_cells &lt;- levels(TME_New$Celltype)[as.numeric(names(sort(rs)[1:5]))]
## }
## debug: ciber_theme &lt;- ggplot2::theme(plot.title = element_text(size = 12, 
##     color = &quot;black&quot;, hjust = 0.5), axis.title = element_text(size = 10, 
##     color = &quot;black&quot;), axis.text = element_text(size = 10, color = &quot;black&quot;), 
##     axis.text.x = element_text(angle = 45, hjust = 1), legend.position = &quot;top&quot;, 
##     legend.text = element_text(size = 12), legend.title = element_text(size = 12))
## debug: if (style == &quot;raw&quot;) {
##     box_TME &lt;- ggplot2::ggplot(TME_New[TME_New$Celltype %in% 
##         selected_cells, ], aes(x = .data$Celltype, y = .data$Composition)) + 
##         ggplot2::labs(y = &quot;Cell composition&quot;, x = NULL, title = &quot;TME Cell composition&quot;) + 
##         ggplot2::geom_boxplot(aes(fill = .data$Group), position = position_dodge(0.5), 
##             width = 0.5, outlier.alpha = 0) + ggplot2::scale_fill_manual(values = group_color) + 
##         ggplot2::theme_classic() + ciber_theme
##     y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
##     box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##         label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = TRUE, 
##         label.y = y_max * 1.1) + coord_cartesian(ylim = c(0, 
##         y_max * 1.1))
## }
## debug: if (style == &quot;elegant&quot;) {
##     box_TME &lt;- ggplot(TME_New, aes(x = .data$Celltype, y = .data$Composition, 
##         fill = .data$Group)) + stat_boxplot(data = TME_New, geom = &quot;errorbar&quot;, 
##         width = 1, color = &quot;black&quot;, linetype = &quot;solid&quot;, position = position_dodge(0.8), 
##         linewidth = 0.7) + stat_boxplot(geom = &quot;boxplot&quot;, color = &quot;black&quot;, 
##         linetype = &quot;solid&quot;, position = position_dodge(0.8), linewidth = 0.7, 
##         width = 0.8, outlier.shape = 19) + ggpubr::theme_classic2() + 
##         ciber_theme + scale_fill_manual(values = group_color)
##     y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
##     box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##         label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = FALSE, 
##         label.y = y_max * 1.1, size = 3) + coord_cartesian(ylim = c(0, 
##         y_max * 1.1))
## }
## debug: box_TME &lt;- ggplot(TME_New, aes(x = .data$Celltype, y = .data$Composition, 
##     fill = .data$Group)) + stat_boxplot(data = TME_New, geom = &quot;errorbar&quot;, 
##     width = 1, color = &quot;black&quot;, linetype = &quot;solid&quot;, position = position_dodge(0.8), 
##     linewidth = 0.7) + stat_boxplot(geom = &quot;boxplot&quot;, color = &quot;black&quot;, 
##     linetype = &quot;solid&quot;, position = position_dodge(0.8), linewidth = 0.7, 
##     width = 0.8, outlier.shape = 19) + ggpubr::theme_classic2() + 
##     ciber_theme + scale_fill_manual(values = group_color)
## debug: y_max &lt;- max(ggplot_build(box_TME)$data[[1]]$ymax)
## debug: box_TME &lt;- box_TME + ggpubr::stat_compare_means(ggplot2::aes(group = .data$Group), 
##     label = &quot;p.signif&quot;, method = &quot;wilcox.test&quot;, hide.ns = FALSE, 
##     label.y = y_max * 1.1, size = 3) + coord_cartesian(ylim = c(0, 
##     y_max * 1.1))
## debug: list(t(result), box_TME)</code></pre>
</div>
<div id="more-details-about-tme-analysis" class="section level2 unnumbered hasAnchor">
<h2>üìù More Details About TME Analysis<a href="#more-details-about-tme-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p>‚ÄÉ<strong>TIMER</strong> is a comprehensive tool for systematical analysis of immune infiltrates across diverse cancer types.</p>
<p>‚ÄÉ<strong>CIBERSORT</strong> is an analytical tool from the Alizadeh Lab and Newman Lab to impute gene expression profiles and provide an estimation of the abundances of member cell types in a mixed cell population, using gene expression data.</p>
<p>‚ÄÉ<strong>MCPCounter</strong> is called the Microenvironment Cell Populations-counter. It allows the robust quantification of the absolute abundance of eight immune and two stromal cell populations in heterogeneous tissues from transcriptomic data</p>
<p>‚ÄÉ<strong>xCell</strong> is a gene signature-based method learned from thousands of pure cell types from various sources. xCell applies a novel technique for reducing associations between closely related cell types.</p>
<p>‚ÄÉ<strong>IPS</strong> uses an analytical strategy to provide comprehensive view of 28 TIL subpopulations including effector and memory T cells and immunosuppressive cells (Tregs, MDSCs).</p>
<p>‚ÄÉ<strong>EPIC</strong> is called Estimating the Proportion of Immune and Cancer cells. It compares the level of expression of genes in a tumor with a library of the gene expression profiles from specific cell types that can be found in tumors and uses this information to predict how many of each type of cell are present.</p>
<p>‚ÄÉ<strong>ESTIMATE</strong> is described as ‚ÄòEstimation of STromal and Immune cells in MAlignant Tumours using Expression data‚Äô. It is a method that utilizes gene expression signatures to infer the proportion of stromal and immune cells within tumor samples.</p>
<p>‚ÄÉ<strong>ABIS</strong> is called ABsolute Immune Signal (ABIS) deconvolution. ABIS performs absolute deconvolution on RNA-Seq and microarray data.</p>
<p>‚ÄÉ<strong>Consensus</strong><sup>TME</sup> uses a consensus approch to generate cancer-specific signatres for TME cell types, and utilizes ssGSEA framework to estimate the relative abundance of these cell types.</p>
<p>‚ÄÉ<strong>quanTIseq</strong> is a computational pipeline for the quantification of the Tumor Immune contexture from human RNA-seq data.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="biomarker-evaluation.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="prediction-model-construction.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-TME-Analysis.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
