---
title: "Tumor immunolotherapy gene expression R package"
author: 
- name: Guangchuang Yu
  email: guangchuangyu@gmail.com
  affiliation: Department of Bioinformatics, School of Basic Medical Sciences, Southern Medical University
output:
  prettydoc::html_pretty:
    toc: true
    theme: cayman
    highlight: github


vignette: >
  %\VignetteIndexEntry{Tumor immunolotherapy gene expression R packag}
  %\usepackage[utf8]{inputenc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1 Dataload loading Module

## 1.1 Dataloader()

### 1.1.1 function
&emsp;You can use **Dataloader()** function to load built-in data of tigeR package from ExperimentHub or TIGER web server.
&emsp;Additionally, use **Dataloader()** with no parameter can get a information table of data set information.

### 1.1.2 instance
```
Dataloader()[1:3,1:5]
#           Dataset              Cancer type     PMID         Therapy Dataset Size
# 1 GBM_PRJNA482620             Glioblastoma 30996326       anti-PD-1           34
# 2   HNSC_GSE93157   Head and Neck Squamous 28487385       anti-PD-1            5
# 3 LGG_E_MTAB_6270 Brain Lower Grade Glioma 29618666 Peptide-vaccine           12
```

# 2 Signature Module

## 2.1 Signature_calculation()

### 2.1.1 function

 By employing the Signature_calculation() function, you can obtain a comprehensive signature score matrix for the 23 signatures in TigeR. In this matrix, the columns represent the signature scores, and the rows denote the sample names.

### 2.1.2 algorithm

 The Signature scores are mostly represented by the average mean or weighted mean or ZScore of the expression value of specified genes.  
 For example, MSKCC Signature are calculated by the following formula:  
   
 &emsp;$MSKCC=-0.492\times exp(TP53)+0.562\times exp(PIK3CA)+1.454\times exp(ATM)$  
   
 where exp() represents the expression value of the gene.

### 2.1.3 instance

```         
Signature_calculation(assay(MEL_GSE78220))[1:4, 1:6]

#             IRS       tGE8     MEMTS  PRGScore Angiogenesis Teffector
#SRR3184279 1.142 -0.1192972 11.940531 11.771754    13.484165  3.162074
#SRR3184280 0.563 -0.2286068  2.942161 15.670362     4.718858  2.485058
#SRR3184281 0.002 -0.2820309  7.624309  6.816627     8.210066  2.318390
#SRR3184282 0.673 -0.6163398 12.379746 12.612995     3.466574  1.397927
```

 Collumns represent Signatures and rows represent sample.

## 2.2 Signature_assessment()

### 2.2.1 function
&emsp;By employing the Signature_assessment() function, you can assess the performance of Signature(including user-built Signature) in different datasets. The function will return a "roc"" object, a list of class "roc".

### 2.2.2 parameter
 ***SE*** a SummarizedExperiment object or a list contains SummarizedExperiment objects (including expression profile and patient clinical information).

 ***Signature*** a subset of genes which you want to assess.

 ***rmBE*** a logical variable. If TRUE, the function will remove the batch effect between datasets through the built-in batch effect removal tool *Combat*.

 ***response_NR*** a logical variable.If TRUE, the function will automatically convert the patient's drug response (such as PR, NR, SD, etc. to binary value NR (non-responder) or R (Responder)).

### 2.2.3 instance
```
result <- Signature_assessment(MEL_PRJEB23709,
                               Weighted_mean_Sigs$Tcell_inflamed_GEP,
                               rmBE=TRUE,
                               response_NR=TRUE)
```

# 3 Machine learning Module

## 3.1 build_Model()

### 3.1.1 function

 You can use function **build_Model** to build machine learning predictive model including naive Bayes models, Support vector machine model, random forest model, Cancerclass model, Adaboost model, Logitboost model, Logistics regression model.  The function returns a trained model. 

### 3.1.2 parameter
 ***Model*** represents the type of model you want to build. You have several options to choose from: "NB" for Naive Bayes, "SVM" for Support Vector Machine, "RF" for Random Forest, "CC" for Cancerclass, "ADB" for Adaboost, "LGB" for Logitboost, and "LGT" for Logistics.

 ***SE*** the dataset you wish to use to build your model. A SummarizedExperiment (SE) object, which can be either a single SE object or a list of SE objects. Note that for each SE object, the colData must contain treatment information under the column name Treatment.

 ***feature_gene*** refers to the specific set of genes you wish to use for model construction.

 ***rmBE*** is a logical variable. If TRUE, the function will remove the batch effect between datasets through the built-in batch effect removal tool *Combat*.

 ***response_NR*** is a logical variable.If TRUE, the function will automatically convert the patient's drug response (such as PR, NR, SD, etc. to binary value NR (non-responder) or R (Responder)).

### 3.1.3 instance
```
train_set <- list(MEL_GSE91061, MEL_phs000452, RCC_Braun_2020)
mymodel <- build_Model(Model='NB', SE=train_set, feature_genes=Stem.Sig, response_NR = TRUE)
```

## 3.2 test_Model()
### 3.2.1 function
 **build_Model** are designed for testing model built by **build_Model** function The function returns a ROC object.
### 3.2.2 parameter
 ***Model*** Machine learning model generated by build_Model() function.
 
 ***SE*** an SummarizedExperiment(SE) object or a list consists of SE objects. The colData of SE objects must contain response information.
## 3.2.3 instance

See details in <https:://github/YuLab-SMU/tigeR>

# 4 Immunotherapy Response

## 4.1 Immunotherapy_Response()

### 4.1.1 function

 You can use **Immunotherapy_Response()** function to erform differential expression analysis or survival analysis.  Immunotherapy_Response() perform differential expression analysis and survival analysis in certain gene and return the result.

### 4.1.2 algorithm

 The score of differential expression analysis is calculated by following formula:

  $-SIGN(log2(FC)) \times log10(p)$

 where $FC$ represents the fold change and $p$ represents the P value derived from the Wilcoxon rank-sum test

 The score of survival analysis is calculated by the following formula:

  $-SIGN(log2(HR)) \times log10(p)$

 where $HR$ represents the hazard ratio and $p$ represents the P value derived from univariate Cox regression analysis.

### 4.1.3 parameter

 ***gene*** is the gene which you wanted.
 ***SE*** is a SummarizedExperiment(SE) which contain expression matrix and clinical information.

### 4.1.4 instance

```         
Immunotherapy_Response(SE = MEL_GSE78220)
```

## 4.2 plt_diff()

### 4.2.1 function

 You can use plt_diff() to visualize differential analysis result between Pre-Treatment and Post-Treatment patients or Responders and Non-Responders in specified gene.

### 4.2.2 parameter

 ***gene*** is the Gene or Gene set you are interested in. 
 
 ***SE*** is a SummarizedExperiment(SE) object or a list consists of SE objects. The colData of SE objects must contain response information. 
 
 ***type*** is the type of analysis you want to perform(Responder vs Non-Responder or Pre-Treatment vs Post-Treatment). c('Treatment','Response')

### 4.2.3 instance

***Pre-Treament vs Post-Treatment***

```         
plt_diff('CD274',MEL_GSE91061,'Treatment') +
  ggtitle("Treatment vs UnTreatment") +
  theme(plot.title = element_text(hjust = 0.5)) 
```

***Responder vs Non-Responder***

```         
plt_diff('CD274',MEL_GSE91061,'Response') +
  ggtitle("Responder vs Non-Responder") +
  theme(plot.title = element_text(hjust = 0.5))
```

## 4.3 plt_surv()

### 4.3.1 function

 You can use plt_diff() to visualize survival analysis result in specified gene.

### 4.3.2 parameter

 ***gene*** is the Gene or Gene set you are interested in.

 ***SE*** is a SummarizedExperiment(SE) object or a list consists of SE objects. The colData of SE objects must contain response information.

### 4.3.3 instance

```         
P <- plt_surv('CD274',MEL_GSE91061)
P$plot <- P$plot +
  ggtitle("Survival analysis") +
  theme(plot.title = element_text(hjust = 0.5))
P
```

# 5 Tumor Microenviroment Deconvolution

## 5.1 CIBERSORT()

### 5.1.1 function

 CIBERSORT is an analytical tool from the Alizadeh Lab and Newman Lab to impute gene expression profiles and provide an estimation of the abundances of member cell types in a mixed cell population, using gene expression data.  tigeR offer you an built-in function **CIBERSORT()** for estimation of the abundances of member cell types. **CIBERSORT()** function will return a list which first element is the cell type proportion matrix and second element is a boxplot.

### 5.1.2 algorithm

 See details in

### 5.1.3 parameter

 ***sig_matrix*** is a expression matrix of Signature genes from each type of immune cells LM22 is a matrix which contain expression of Signature genes of each type of immune cell

 ***mix_matrix*** is the expression matrix of bulk RNA-seq which you want to know its cell fraction

 ***perm*** is the number of permutations

 ***QN*** perform quantile normalization or not (TRUE/FALSE)

### 5.1.4 instance

```         
result <- CIBERSORT(sig_matrix = LM22, SE = MEL_GSE78220, perm = 0, QN = TRUE)
```

## 5.2 xCell

## 5.3 ConsensusTME

## 5.4 epic

## 5.5 ABIS

## 5.6 ESTIMATE

## 5.7 quanTIseq

## 5.8 MCPCounter


