<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 BioMarker Evaluation | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</title>
  <meta name="description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 BioMarker Evaluation | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 BioMarker Evaluation | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  
  <meta name="twitter:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  

<meta name="author" content="Yihao Chen1,2,4,#, Lina He2,#, Yuanzhe Zhang2, Yuelong Shu1, Di Zhang3,, Guangchuang Yu4,, Zhixiang Zuo2,*" />


<meta name="date" content="2024-05-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="load-built-in-data-into-r-environment.html"/>
<link rel="next" href="tumor-microenviroment-analysis.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Getting Started with tigeR</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#introuction"><i class="fa fa-check"></i><b>1.1</b> Introuction</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.2</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html"><i class="fa fa-check"></i><b>2</b> Load built-in data into R environment</a>
<ul>
<li class="chapter" data-level="2.1" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-data-packagerecommandate"><i class="fa fa-check"></i><b>2.1</b> Obtain data from data package(recommandate)</a></li>
<li class="chapter" data-level="2.2" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-tiger-web-server"><i class="fa fa-check"></i><b>2.2</b> Obtain data from tigeR web server</a></li>
<li class="chapter" data-level="2.3" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-experimenthub"><i class="fa fa-check"></i><b>2.3</b> Obtain data from ExperimentHub</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html"><i class="fa fa-check"></i><b>3</b> BioMarker Evaluation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#integrate-analysis"><i class="fa fa-check"></i><b>3.1</b> Integrate analysis</a></li>
<li class="chapter" data-level="3.2" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#differential-analysis"><i class="fa fa-check"></i><b>3.2</b> Differential analysis</a></li>
<li class="chapter" data-level="3.3" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#suvival-analysis"><i class="fa fa-check"></i><b>3.3</b> Suvival analysis</a></li>
<li class="chapter" data-level="3.4" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#calculate-comprehensive-signature-score"><i class="fa fa-check"></i><b>3.4</b> Calculate comprehensive signature score</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#algorithm"><i class="fa fa-check"></i><b>3.4.1</b> algorithm</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#assess-the-performance-of-signature"><i class="fa fa-check"></i><b>3.5</b> Assess the performance of Signature</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html"><i class="fa fa-check"></i><b>4</b> Tumor Microenviroment Analysis</a>
<ul>
<li class="chapter" data-level="4.1" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#timer"><i class="fa fa-check"></i><b>4.1</b> TIMER</a></li>
<li class="chapter" data-level="4.2" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#cibersort"><i class="fa fa-check"></i><b>4.2</b> CIBERSORT</a></li>
<li class="chapter" data-level="4.3" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#xcell"><i class="fa fa-check"></i><b>4.3</b> 4.3 xCell</a></li>
<li class="chapter" data-level="4.4" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#consensustme"><i class="fa fa-check"></i><b>4.4</b> 4.4 ConsensusTME</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html"><i class="fa fa-check"></i><b>5</b> Prediction Model Construction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#build_model"><i class="fa fa-check"></i><b>5.1</b> build_Model</a></li>
<li class="chapter" data-level="5.2" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#test_model"><i class="fa fa-check"></i><b>5.2</b> test_Model</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="response-prediction.html"><a href="response-prediction.html"><i class="fa fa-check"></i><b>6</b> Response Prediction</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="biomarker-evaluation" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">Chapter 3</span> BioMarker Evaluation<a href="biomarker-evaluation.html#biomarker-evaluation" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="integrate-analysis" class="section level2 hasAnchor" number="3.1">
<h2><span class="header-section-number">3.1</span> Integrate analysis<a href="biomarker-evaluation.html#integrate-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>integrate_analysis(SE=MEL_GSE91061, geneSet=&quot;CD274&quot;)</code></pre>
</div>
<div id="differential-analysis" class="section level2 hasAnchor" number="3.2">
<h2><span class="header-section-number">3.2</span> Differential analysis<a href="biomarker-evaluation.html#differential-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p> You can use diff_biomk() to visualize differential analysis result between Pre-Treatment and Post-Treatment patients or Responders and Non-Responders in specified gene.</p>
<p><strong><em>Pre-Treament vs Post-Treatment</em></strong></p>
<pre><code>diff_biomk(SE=MEL_GSE91061,gene=&#39;CD274&#39;,type=&#39;Treatment&#39;) +
  ggtitle(&quot;Treatment vs UnTreatment&quot;) +
  theme(plot.title = element_text(hjust = 0.5)) </code></pre>
<p><strong><em>Responder vs Non-Responder</em></strong></p>
<pre><code>diff_biomk(SE=MEL_GSE91061,gene=&#39;CD274&#39;,type=&#39;Response&#39;) +
  ggtitle(&quot;Responder vs Non-Responder&quot;) +
  theme(plot.title = element_text(hjust = 0.5))</code></pre>
</div>
<div id="suvival-analysis" class="section level2 hasAnchor" number="3.3">
<h2><span class="header-section-number">3.3</span> Suvival analysis<a href="biomarker-evaluation.html#suvival-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p> You can use diff_biomk() to visualize survival analysis result in specified gene.</p>
<pre><code>P &lt;- surv_biomk(SE=MEL_GSE91061,gene=&#39;CD274&#39;)
P$plot &lt;- P$plot +
  ggtitle(&quot;Survival analysis&quot;) +
  theme(plot.title = element_text(hjust = 0.5))
P</code></pre>
</div>
<div id="calculate-comprehensive-signature-score" class="section level2 hasAnchor" number="3.4">
<h2><span class="header-section-number">3.4</span> Calculate comprehensive signature score<a href="biomarker-evaluation.html#calculate-comprehensive-signature-score" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p> By employing the score_biomk() function, you can obtain a comprehensive signature score matrix for the 23 signatures in TigeR.
In this matrix, the columns represent the signature scores, and the rows denote the sample names.</p>
<table>
<colgroup>
<col width="24%" />
<col width="24%" />
<col width="24%" />
<col width="28%" />
</colgroup>
<thead>
<tr class="header">
<th>Signature</th>
<th>Method</th>
<th>Citation</th>
<th>PMID</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>IRS</td>
<td>multivariate Cox analysis</td>
<td>Zhou R, Liang J, Tian H, Chen Q, Yang C, Liu C. An Immunosenescence-Related Gene Signature to Evaluate the Prognosis, Immunotherapeutic Response, and Cisplatin Sensitivity of Bladder Cancer. Dis Markers. 2022 Mar 2;2022:2143892. doi: 10.1155/2022/2143892. PMID: 35280438; PMCID: PMC8915927.</td>
<td>35280438</td>
</tr>
<tr class="even">
<td>tGE8</td>
<td>median of Z-score</td>
<td>Powles T, Kockx M, Rodriguez-Vida A, Duran I, Crabb SJ, Van Der Heijden MS, Szabados B, Pous AF, Gravis G, Herranz UA, Protheroe A, Ravaud A, Maillet D, Mendez MJ, Suarez C, Linch M, Prendergast A, van Dam PJ, Stanoeva D, Daelemans S, Mariathasan S, Tea JS, Mousa K, Banchereau R, Castellano D. Clinical efficacy and biomarker analysis of neoadjuvant atezolizumab in operable urothelial carcinoma in the ABACUS trial. Nat Med. 2019 Nov;25(11):1706-1714. doi: 10.1038/s41591-019-0628-7. Epub 2019 Nov 4. Erratum in: Nat Med. 2020 Jun;26(6):983. Erratum in: Nat Med. 2023 Dec;29(12):3271. PMID: 31686036.</td>
<td>31686036</td>
</tr>
<tr class="odd">
<td>MEMTS</td>
<td>Average mean</td>
<td>Song J, Wei R, Huo S, Gao J, Liu X. Metastasis Related Epithelial-Mesenchymal Transition Signature Predicts Prognosis and Response to Immunotherapy in Gastric Cancer. Front Immunol. 2022 Jun 13;13:920512. doi: 10.3389/fimmu.2022.920512. PMID: 35769483; PMCID: PMC9234207.</td>
<td>35769483</td>
</tr>
<tr class="even">
<td>PRGScore</td>
<td>Average mean</td>
<td>Zhang Q, Tan Y, Zhang J, Shi Y, Qi J, Zou D, Ci W. Pyroptosis-Related Signature Predicts Prognosis and Immunotherapy Efficacy in Muscle-Invasive Bladder Cancer. Front Immunol. 2022 Apr 11;13:782982. doi: 10.3389/fimmu.2022.782982. PMID: 35479097; PMCID: PMC9035667.</td>
<td>35479097</td>
</tr>
<tr class="odd">
<td>Angiogenesis</td>
<td>Average mean</td>
<td>McDermott DF, Huseni MA, Atkins MB, Motzer RJ, Rini BI, Escudier B, Fong L, Joseph RW, Pal SK, Reeves JA, Sznol M, Hainsworth J, Rathmell WK, Stadler WM, Hutson T, Gore ME, Ravaud A, Bracarda S, Suárez C, Danielli R, Gruenwald V, Choueiri TK, Nickles D, Jhunjhunwala S, Piault-Louis E, Thobhani A, Qiu J, Chen DS, Hegde PS, Schiff C, Fine GD, Powles T. Clinical activity and molecular correlates of response to atezolizumab alone or in combination with bevacizumab versus sunitinib in renal cell carcinoma. Nat Med. 2018 Jun;24(6):749-757. doi: 10.1038/s41591-018-0053-3. Epub 2018 Jun 4. Erratum in: Nat Med. 2018 Dec;24(12):1941. PMID: 29867230; PMCID: PMC6721896.</td>
<td>29867230</td>
</tr>
<tr class="even">
<td>Teffector</td>
<td>Average mean</td>
<td>McDermott DF, Huseni MA, Atkins MB, Motzer RJ, Rini BI, Escudier B, Fong L, Joseph RW, Pal SK, Reeves JA, Sznol M, Hainsworth J, Rathmell WK, Stadler WM, Hutson T, Gore ME, Ravaud A, Bracarda S, Suárez C, Danielli R, Gruenwald V, Choueiri TK, Nickles D, Jhunjhunwala S, Piault-Louis E, Thobhani A, Qiu J, Chen DS, Hegde PS, Schiff C, Fine GD, Powles T. Clinical activity and molecular correlates of response to atezolizumab alone or in combination with bevacizumab versus sunitinib in renal cell carcinoma. Nat Med. 2018 Jun;24(6):749-757. doi: 10.1038/s41591-018-0053-3. Epub 2018 Jun 4. Erratum in: Nat Med. 2018 Dec;24(12):1941. PMID: 29867230; PMCID: PMC6721896.</td>
<td>29867230</td>
</tr>
<tr class="odd">
<td>Myeloid_inflammatory</td>
<td>Average mean</td>
<td>McDermott DF, Huseni MA, Atkins MB, Motzer RJ, Rini BI, Escudier B, Fong L, Joseph RW, Pal SK, Reeves JA, Sznol M, Hainsworth J, Rathmell WK, Stadler WM, Hutson T, Gore ME, Ravaud A, Bracarda S, Suárez C, Danielli R, Gruenwald V, Choueiri TK, Nickles D, Jhunjhunwala S, Piault-Louis E, Thobhani A, Qiu J, Chen DS, Hegde PS, Schiff C, Fine GD, Powles T. Clinical activity and molecular correlates of response to atezolizumab alone or in combination with bevacizumab versus sunitinib in renal cell carcinoma. Nat Med. 2018 Jun;24(6):749-757. doi: 10.1038/s41591-018-0053-3. Epub 2018 Jun 4. Erratum in: Nat Med. 2018 Dec;24(12):1941. PMID: 29867230; PMCID: PMC6721896.</td>
<td>29867230</td>
</tr>
<tr class="even">
<td>IFNG_Sig</td>
<td>Average mean</td>
<td>Mo X, Zhang H, Preston S, Martin K, Zhou B, Vadalia N, Gamero AM, Soboloff J, Tempera I, Zaidi MR. Interferon-γ Signaling in Melanocytes and Melanoma Cells Regulates Expression of CTLA-4. Cancer Res. 2018 Jan 15;78(2):436-450. doi: 10.1158/0008-5472.CAN-17-1615. Epub 2017 Nov 17. PMID: 29150430; PMCID: PMC5771950.</td>
<td>29150430</td>
</tr>
<tr class="odd">
<td>TLS</td>
<td>Weighted mean</td>
<td>Cabrita R, Lauss M, Sanna A, Donia M, Skaarup Larsen M, Mitra S, Johansson I, Phung B, Harbst K, Vallon-Christersson J, van Schoiack A, Lövgren K, Warren S, Jirström K, Olsson H, Pietras K, Ingvar C, Isaksson K, Schadendorf D, Schmidt H, Bastholt L, Carneiro A, Wargo JA, Svane IM, Jönsson G. Tertiary lymphoid structures improve immunotherapy and survival in melanoma. Nature. 2020 Jan;577(7791):561-565. doi: 10.1038/s41586-019-1914-8. Epub 2020 Jan 15. Erratum in: Nature. 2020 Apr;580(7801):E1. PMID: 31942071.</td>
<td>31942071</td>
</tr>
<tr class="even">
<td>MSKCC</td>
<td>Weighted mean</td>
<td>Pan YH, Zhang JX, Chen X, Liu F, Cao JZ, Chen Y, Chen W, Luo JH. Predictive Value of the TP53/PIK3CA/ATM Mutation Classifier for Patients With Bladder Cancer Responding to Immune Checkpoint Inhibitor Therapy. Front Immunol. 2021 Aug 4;12:643282. doi: 10.3389/fimmu.2021.643282. PMID: 34421886; PMCID: PMC8371040.</td>
<td>34421886</td>
</tr>
<tr class="odd">
<td>LMRGPI</td>
<td>Weighted mean</td>
<td>Jiang A, Chen X, Zheng H, Liu N, Ding Q, Li Y, Fan C, Fu X, Liang X, Tian T, Ruan Z, Yao Y. Lipid metabolism-related gene prognostic index (LMRGPI) reveals distinct prognosis and treatment patterns for patients with early-stage pulmonary adenocarcinoma. Int J Med Sci. 2022 Mar 28;19(4):711-728. doi: 10.7150/ijms.71267. PMID: 35582412; PMCID: PMC9108406.</td>
<td>35582412</td>
</tr>
<tr class="even">
<td>PRS</td>
<td>Weighted mean</td>
<td>Yu H, Fu Y, Tang Z, Jiang L, Qu C, Li H, Tan Z, Shu D, Peng Y, Liu S. A novel pyroptosis-related signature predicts prognosis and response to treatment in breast carcinoma. Aging (Albany NY). 2022 Jan 27;14(2):989-1013. doi: 10.18632/aging.203855. Epub 2022 Jan 27. PMID: 35085103; PMCID: PMC8833126.</td>
<td>35085103</td>
</tr>
<tr class="odd">
<td>Stemnesssignatures</td>
<td>Weighted mean</td>
<td>Zheng H, Liu H, Li H, Dou W, Wang J, Zhang J, Liu T, Wu Y, Liu Y, Wang X. Characterization of stem cell landscape and identification of stemness-relevant prognostic gene signature to aid immunotherapy in colorectal cancer. Stem Cell Res Ther. 2022 Jun 9;13(1):244. doi: 10.1186/s13287-022-02913-0. PMID: 35681225; PMCID: PMC9185878.</td>
<td>35681225</td>
</tr>
<tr class="even">
<td>GRIP</td>
<td>Weighted mean</td>
<td>Xu Y, Chen Y, Niu Z, Xing J, Yang Z, Yin X, Guo L, Zhang Q, Qiu H, Han Y. A Novel Pyroptotic and Inflammatory Gene Signature Predicts the Prognosis of Cutaneous Melanoma and the Effect of Anticancer Therapies. Front Med (Lausanne). 2022 Apr 15;9:841568. doi: 10.3389/fmed.2022.841568. PMID: 35492358; PMCID: PMC9053829.</td>
<td>35492358</td>
</tr>
<tr class="odd">
<td>IPS</td>
<td>Weighted mean</td>
<td>Zhao B, Wang Y, Wang Y, Chen W, Liu PH, Kong Z, Dai C, Wang Y, Ma W. Systematic identification, development, and validation of prognostic biomarkers involving the tumor-immune microenvironment for glioblastoma. J Cell Physiol. 2021 Jan;236(1):507-522. doi: 10.1002/jcp.29878. Epub 2020 Jun 22. PMID: 32572951.</td>
<td>32572951</td>
</tr>
<tr class="even">
<td>Tcell_inflamed_GEP</td>
<td>Weighted mean</td>
<td>Cristescu R, Mogg R, Ayers M, Albright A, Murphy E, Yearley J, Sher X, Liu XQ, Lu H, Nebozhyn M, Zhang C, Lunceford JK, Joe A, Cheng J, Webber AL, Ibrahim N, Plimack ER, Ott PA, Seiwert TY, Ribas A, McClanahan TK, Tomassini JE, Loboda A, Kaufman D. Pan-tumor genomic biomarkers for PD-1 checkpoint blockade-based immunotherapy. Science. 2018 Oct 12;362(6411):eaar3593. doi: 10.1126/science.aar3593. Erratum in: Science. 2019 Mar 1;363(6430): PMID: 30309915; PMCID: PMC6718162.</td>
<td>30309915</td>
</tr>
<tr class="odd">
<td>DDR</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="even">
<td>CD8Teffector</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="odd">
<td>CellCycleReg</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="even">
<td>PanFTBRs</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="odd">
<td>EMT1</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="even">
<td>EMT2</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
<tr class="odd">
<td>EMT3</td>
<td>Z-score;PCA</td>
<td>Mariathasan S, Turley SJ, Nickles D, Castiglioni A, Yuen K, Wang Y, Kadel EE III, Koeppen H, Astarita JL, Cubas R, Jhunjhunwala S, Banchereau R, Yang Y, Guan Y, Chalouni C, Ziai J, Şenbabaoğlu Y, Santoro S, Sheinson D, Hung J, Giltnane JM, Pierce AA, Mesh K, Lianoglou S, Riegler J, Carano RAD, Eriksson P, Höglund M, Somarriba L, Halligan DL, van der Heijden MS, Loriot Y, Rosenberg JE, Fong L, Mellman I, Chen DS, Green M, Derleth C, Fine GD, Hegde PS, Bourgon R, Powles T. TGFβ attenuates tumour response to PD-L1 blockade by contributing to exclusion of T cells. Nature. 2018 Feb 22;554(7693):544-548. doi: 10.1038/nature25501. Epub 2018 Feb 14. PMID: 29443960; PMCID: PMC6028240.</td>
<td>29443960</td>
</tr>
</tbody>
</table>
<div id="algorithm" class="section level3 hasAnchor" number="3.4.1">
<h3><span class="header-section-number">3.4.1</span> algorithm<a href="biomarker-evaluation.html#algorithm" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<p> The Signature scores are mostly represented by the average mean or weighted mean or ZScore of the expression value of specified genes.<br />
 For example, MSKCC Signature are calculated by the following formula:<br />
 <br />
  <span class="math inline">\(MSKCC=-0.492\times exp(TP53)+0.562\times exp(PIK3CA)+1.454\times exp(ATM)\)</span><br />
 <br />
 where exp() represents the expression value of the gene.</p>
<pre><code>score_biomk(assay(MEL_GSE78220))[1:4, 1:6]

#             IRS       tGE8     MEMTS  PRGScore Angiogenesis Teffector
#SRR3184279 1.142 -0.1192972 11.940531 11.771754    13.484165  3.162074
#SRR3184280 0.563 -0.2286068  2.942161 15.670362     4.718858  2.485058
#SRR3184281 0.002 -0.2820309  7.624309  6.816627     8.210066  2.318390
#SRR3184282 0.673 -0.6163398 12.379746 12.612995     3.466574  1.397927</code></pre>
<p> Columns represent Signatures and rows represent sample.</p>
</div>
</div>
<div id="assess-the-performance-of-signature" class="section level2 hasAnchor" number="3.5">
<h2><span class="header-section-number">3.5</span> Assess the performance of Signature<a href="biomarker-evaluation.html#assess-the-performance-of-signature" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<p> By employing the auc_biomk() function, you can assess the performance of Signature(including user-built Signature) in different datasets.
The function will return a “roc”” object, a list of class “roc”.</p>
<pre><code>result &lt;- auc_biomk(MEL_PRJEB23709,
                               Weighted_mean_Sigs$Tcell_inflamed_GEP,
                               rmBE=TRUE,
                               response_NR=TRUE)</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="load-built-in-data-into-r-environment.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="tumor-microenviroment-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/02-cross-refs.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
