<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 2 Load built-in data into R environment | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</title>
  <meta name="description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 2 Load built-in data into R environment | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  <meta name="github-repo" content="rstudio/bookdown-demo" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 2 Load built-in data into R environment | tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package" />
  
  <meta name="twitter:description" content="<p>This is a minimal example of using the bookdown package to write a book.
The HTML output format for this example is bookdown::gitbook,
set in the _output.yml file.</p>" />
  

<meta name="author" content="Yihao Chen1,2,4,#, Lina He2,#, Yuanzhe Zhang2, Yuelong Shu1, Di Zhang3,, Guangchuang Yu4,, Zhixiang Zuo2,*" />


<meta name="date" content="2024-05-09" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="index.html"/>
<link rel="next" href="biomarker-evaluation.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Getting Started with tigeR</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#introuction"><i class="fa fa-check"></i><b>1.1</b> Introuction</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#installation"><i class="fa fa-check"></i><b>1.2</b> Installation</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html"><i class="fa fa-check"></i><b>2</b> Load built-in data into R environment</a>
<ul>
<li class="chapter" data-level="2.1" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-data-packagerecommandate"><i class="fa fa-check"></i><b>2.1</b> Obtain data from data package(recommandate)</a></li>
<li class="chapter" data-level="2.2" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-tiger-web-server"><i class="fa fa-check"></i><b>2.2</b> Obtain data from tigeR web server</a></li>
<li class="chapter" data-level="2.3" data-path="load-built-in-data-into-r-environment.html"><a href="load-built-in-data-into-r-environment.html#obtain-data-from-experimenthub"><i class="fa fa-check"></i><b>2.3</b> Obtain data from ExperimentHub</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html"><i class="fa fa-check"></i><b>3</b> BioMarker Evaluation</a>
<ul>
<li class="chapter" data-level="3.1" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#integrate-analysis"><i class="fa fa-check"></i><b>3.1</b> Integrate analysis</a></li>
<li class="chapter" data-level="3.2" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#differential-analysis"><i class="fa fa-check"></i><b>3.2</b> Differential analysis</a></li>
<li class="chapter" data-level="3.3" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#suvival-analysis"><i class="fa fa-check"></i><b>3.3</b> Suvival analysis</a></li>
<li class="chapter" data-level="3.4" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#calculate-comprehensive-signature-score"><i class="fa fa-check"></i><b>3.4</b> Calculate comprehensive signature score</a>
<ul>
<li class="chapter" data-level="3.4.1" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#algorithm"><i class="fa fa-check"></i><b>3.4.1</b> algorithm</a></li>
</ul></li>
<li class="chapter" data-level="3.5" data-path="biomarker-evaluation.html"><a href="biomarker-evaluation.html#assess-the-performance-of-signature"><i class="fa fa-check"></i><b>3.5</b> Assess the performance of Signature</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html"><i class="fa fa-check"></i><b>4</b> Tumor Microenviroment Analysis</a>
<ul>
<li class="chapter" data-level="4.1" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#timer"><i class="fa fa-check"></i><b>4.1</b> TIMER</a></li>
<li class="chapter" data-level="4.2" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#cibersort"><i class="fa fa-check"></i><b>4.2</b> CIBERSORT</a></li>
<li class="chapter" data-level="4.3" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#xcell"><i class="fa fa-check"></i><b>4.3</b> 4.3 xCell</a></li>
<li class="chapter" data-level="4.4" data-path="tumor-microenviroment-analysis.html"><a href="tumor-microenviroment-analysis.html#consensustme"><i class="fa fa-check"></i><b>4.4</b> 4.4 ConsensusTME</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html"><i class="fa fa-check"></i><b>5</b> Prediction Model Construction</a>
<ul>
<li class="chapter" data-level="5.1" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#build_model"><i class="fa fa-check"></i><b>5.1</b> build_Model</a></li>
<li class="chapter" data-level="5.2" data-path="prediction-model-construction.html"><a href="prediction-model-construction.html#test_model"><i class="fa fa-check"></i><b>5.2</b> test_Model</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="response-prediction.html"><a href="response-prediction.html"><i class="fa fa-check"></i><b>6</b> Response Prediction</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">tigeR: Tumor Immunotherapy Gene Expression Data Analysis R package</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="load-built-in-data-into-r-environment" class="section level1 hasAnchor" number="2">
<h1><span class="header-section-number">Chapter 2</span> Load built-in data into R environment<a href="load-built-in-data-into-r-environment.html#load-built-in-data-into-r-environment" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>The built-in data follows several specifications for constructing:</p>
<p>① They are constructed into a SummarizedExperiment object using the SummarizedExperiment() function in SummarizedExperiment package.</p>
<p>② The assays of the SummarizedExperiment object are stored as a SimpleList containing a expression matrix, where rows represent genes and columns represent patients.</p>
<p>③ The colData of the SummarizedExperiment object is a DataFrame, where the row names must be the same as the column names of the expression matrix.</p>
<p>④ The column names of the DataFrame represent the clinical information of patients. The detailed information about the column names is presented below.</p>
<table>
<colgroup>
<col width="17%" />
<col width="17%" />
<col width="47%" />
<col width="17%" />
</colgroup>
<thead>
<tr class="header">
<th align="center">column name</th>
<th align="center">Recommand value</th>
<th align="center">Description</th>
<th align="center">Necessity in tigeR analysis</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="center">sample id</td>
<td align="center">character</td>
<td align="center">The sample identification.</td>
<td align="center">✕</td>
</tr>
<tr class="even">
<td align="center">patient_name</td>
<td align="center">character</td>
<td align="center">The patient identification.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">dataset_id</td>
<td align="center">character</td>
<td align="center">The dataset identification.</td>
<td align="center">✕</td>
</tr>
<tr class="even">
<td align="center">dataset_group</td>
<td align="center">character</td>
<td align="center"></td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">Treatment</td>
<td align="center">PRE/POST</td>
<td align="center">The patient sample is collected after Treatment or before Treatment</td>
<td align="center">✓</td>
</tr>
<tr class="even">
<td align="center">response</td>
<td align="center">CR/MR/PR/PD/SD/NR/R/N</td>
<td align="center">The Immunotherapy response.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">response_NR</td>
<td align="center">R/N</td>
<td align="center">The Immunotherapy response which only contains R or N.</td>
<td align="center">✓</td>
</tr>
<tr class="even">
<td align="center">M.stage</td>
<td align="center">character</td>
<td align="center">The M stage of patients.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">overall.survival..days.</td>
<td align="center">value</td>
<td align="center">The number of days, months or years of survival time of patients.(all samples must be the same unit)</td>
<td align="center">✓</td>
</tr>
<tr class="even">
<td align="center">vital.status</td>
<td align="center">Alive/Dead</td>
<td align="center">The survival status of patient.</td>
<td align="center">✓</td>
</tr>
<tr class="odd">
<td align="center">Total.Mutation</td>
<td align="center">numeric</td>
<td align="center">The total mutation gene numbers.</td>
<td align="center">✕</td>
</tr>
<tr class="even">
<td align="center">Gender</td>
<td align="center">M/F</td>
<td align="center">The gender of patient.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">Therapy</td>
<td align="center">character</td>
<td align="center">The anti-tumor therapy on patients.</td>
<td align="center">✕</td>
</tr>
<tr class="even">
<td align="center">age_start</td>
<td align="center">numeric</td>
<td align="center">When the tumor is diagnosed.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">tumor_type</td>
<td align="center">character</td>
<td align="center">The type of tumor.</td>
<td align="center">✕</td>
</tr>
<tr class="even">
<td align="center">seq_type</td>
<td align="center">character</td>
<td align="center">The sequencing type.</td>
<td align="center">✕</td>
</tr>
<tr class="odd">
<td align="center">id</td>
<td align="center">character</td>
<td align="center">The identification of dataset.</td>
<td align="center">✕</td>
</tr>
</tbody>
</table>
<p>The column names of the DataFrame represent the clinical information of patients. The detailed information about the column names is presented below.</p>
<div id="obtain-data-from-data-packagerecommandate" class="section level2 hasAnchor" number="2.1">
<h2><span class="header-section-number">2.1</span> Obtain data from data package(recommandate)<a href="load-built-in-data-into-r-environment.html#obtain-data-from-data-packagerecommandate" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>devtools::install.github(&quot;YuLab-SMU/tigeR&quot;)</code></pre>
</div>
<div id="obtain-data-from-tiger-web-server" class="section level2 hasAnchor" number="2.2">
<h2><span class="header-section-number">2.2</span> Obtain data from tigeR web server<a href="load-built-in-data-into-r-environment.html#obtain-data-from-tiger-web-server" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>Dataloader(c(1,2,3), use_source=&quot;Web Server&quot;)</code></pre>
</div>
<div id="obtain-data-from-experimenthub" class="section level2 hasAnchor" number="2.3">
<h2><span class="header-section-number">2.3</span> Obtain data from ExperimentHub<a href="load-built-in-data-into-r-environment.html#obtain-data-from-experimenthub" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<pre><code>Dataloader(c(1,2,3), use_source=&quot;ExperimentHub)</code></pre>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="biomarker-evaluation.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/01-intro.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
